src/instana/instrumentation/pep0249.py

```diff
@@ -2,139 +2,161 @@
             try:
-                self._collect_kvs(scope.span, sql)
-
+                self._collect_kvs(span, sql)
                 result = self.__wrapped__.execute(sql, params)
             except Exception as e:
-                if scope.span:
-                    scope.span.log_exception(e)
+                if span:
+                    span.record_exception(e)
                 raise
             else:
                 return result
 
-    def executemany(self, sql, seq_of_parameters):
-        tracer, parent_span, operation_name = get_tracer_tuple()
+    def executemany(self, sql, seq_of_parameters) -> None:
+        tracer, _, operation_name = get_tracer_tuple()
 
         # If not tracing or we're being called from sqlalchemy, just pass through
-        if (tracing_is_off() or (operation_name == "sqlalchemy")):
+        if tracing_is_off() or (operation_name == "sqlalchemy"):
             return self.__wrapped__.executemany(sql, seq_of_parameters)
 
-        with tracer.start_active_span(self._module_name, child_of=parent_span) as scope:
+        with tracer.start_as_current_span(self._module_name) as span:
             try:
-                self._collect_kvs(scope.span, sql)
-
+                self._collect_kvs(span, sql)
                 result = self.__wrapped__.executemany(sql, seq_of_parameters)
             except Exception as e:
-                if scope.span:
-                    scope.span.log_exception(e)
+                if span:
+                    span.record_exception(e)
                 raise
             else:
                 return result
 
-    def callproc(self, proc_name, params):
-        tracer, parent_span, operation_name = get_tracer_tuple()
+    def callproc(self, proc_name, params) -> None:
+        tracer, _, operation_name = get_tracer_tuple()
 
         # If not tracing or we're being called from sqlalchemy, just pass through
-        if (tracing_is_off() or (operation_name == "sqlalchemy")):
+        if tracing_is_off() or (operation_name == "sqlalchemy"):
             return self.__wrapped__.execute(proc_name, params)
 
-        with tracer.start_active_span(self._module_name, child_of=parent_span) as scope:
+        with tracer.start_as_current_span(self._module_name) as span:
             try:
-                self._collect_kvs(scope.span, proc_name)
-
+                self._collect_kvs(span, proc_name)
                 result = self.__wrapped__.callproc(proc_name, params)
-            except Exception as e:
-                if scope.span:
-                    scope.span.log_exception(e)
-                raise
+            except Exception:
+                try:
+                    result = self.__wrapped__.execute(proc_name, params)
+                except Exception as e_execute:
+                    if span:
+                        span.record_exception(e_execute)
+                    raise
+                else:
+                    return result
             else:
                 return result
```

Comment:
The nested try except blocks look a bit clumpsy to me. Why both `execute` and `callproc` are called on `__wrapped__`?

Reply:
All databases support execute but only some of them support callproc for procedure calls e.g. postgres. 
We can call procedures with `call procedure_name();` command in the db but some libraries like psycopg2 doesn't support callproc function for procedures, so we're using execute if callproc fails.

References:
https://www.psycopg.org/docs/cursor.html#cursor.callproc
https://dev.mysql.com/doc/connector-python/en/connector-python-api-mysqlcursor-callproc.html

Reply:
Not sure if this suggestion of mine would work but I think its worth a try. Can you check if the `__wrapped__` instance has the method attribute `callproc` , `execute` and call them only if present?

If this doesn't work, feel free to keep your current changes.

Reply:
__wrapped__ has callproc attribute but it supports only db functions, not procedures.